vignettes_src/rng_distributed.Rmd

```diff
@@ -0,0 +1,336 @@
+      p <- p_new
+    }
+    ret[i + 1, ] <- c(p, ll)
+  }
+  ret
+}
+```
+
+We can run this for a number of steps and collect up sampled parameters and their likelihoods
+
+```{r}
+ans <- mcmc(mod, c(0.91, 1), 20)
+```
+
+We are cutting a lot of corners here; we have not specified any priors and are assuming that the distribution can be integrated over the parameters, and our proposal mechanism is extremely simple supporting only orthogonal proposals of the two parameters which are certainly correlated. However, the basic features are shared with any more sophisticated approach.
+
+Our aim here is now to carry this out in parallel, where we run a chain per node.  To do this we need to do the same basic steps as above. If we want to run 8 chains we'll need 8 seeds, even if we run on fewer nodes:
+
+```{r}
+seed <- dust::dust_rng_distributed_state(n_nodes = 8L, algorithm = volatility)
+```
+
+It will also be useful to write a little wrapper function that repeats the setup from above, then runs the MCMC:
+
+```{r}
+run_mcmc <- function(seed, data, ...) {
+  volatility <- dust::dust_example("volatility")
+  mod <- volatility$new(list(), 0, 128, seed = seed, n_threads = 2)
+  mod$set_data(dust::dust_data(data))
+  mcmc(mod, ...)
+}
+```
+
+```{r}
+parallel::clusterExport(cl, c("mcmc", "run_mcmc", "run_filter"))
+ans <- parallel::clusterApply(cl, seed, run_mcmc, data, c(0.91, 1), 20)
+```
+
+This produces a list of samples from each of the 8 chains, run over 2 nodes, each of which ran using 2 threads.  We can scale any of this parallelism (increasing threads until we hit the total number of particles, increasing nodes until we hit the number of chains) but the results will be deterministic (except for the call to `runif` that we use here for the acceptance test within the MCMC).
+
+Note that in creating the seed above we left `n_streams` as the default (1) because the dust model constructor will take care of expanding out per-particle seeds via a series of jumps.
+
+You should be careful with these approaches to not exceed the compute available to you, especially when using a multi-user system such as an HPC.
+
+## Summary
+
+* *Simplest*: create a set of suitable rng seeds on the controlling node and send them to the worker nodes, don't try and continue calculations afterwards.
+* Create a set of pointer objects which could be used on the controlling process first, then sent to the worker nodes
+* *Hardest*: create a set of pointer objects and send them to the worker nodes, then at the end of the calculation synchronise the state and return to the controlling node
+
+---
+
+[^1] It's hard to get a sense of these numbers but `2^128` is about `3 x 10^38` draws. If we draw a billion numbers (10^9) a second (1ns per draw is the rate on a reasonable CPU) we can draw 3 x 10^16 numbers a year, so at this rate it would take this would take 10^22 years to reach the next stream. In contrast, the universe is 13 billion years old (13 x 10^9 years).
```

Comment:
```suggestion
[^1] It's hard to get a sense of these numbers but `2^128` is about `3 x 10^38` draws. If we draw a billion numbers (10^9) a second (1ns per draw is the rate on a reasonable CPU) we can draw 3 x 10^16 numbers a year, so at this rate it would take this would take 10^22 years to reach the next stream. In contrast, the universe is 14 billion years old (1.4 x 10^10 years).
```